### ISSUE TITLE
Add ignore: ["after-declaration"] option to declaration-empty-line-before

### ISSUE THREAD

**Issue by heatwaveo8** [stylelint/stylelint#1848]
> Describe the issue. Is it a bug or a feature request (new rule, new option, etc.)?

We have the `block-closing-brace-empty-line-before` rule already available to us to detect empty lines at the end of multi-line blocks. I haven't found a related rule to detect empty lines specifically at the top of multi-line code blocks.

For this, we'd detect the first newline of the block and, if not allowed, throw an error if the next line is empty.

> What CSS is needed to reproduce this issue?

As it pertains to this request, the following would be valid:

```css
.navbar {
    border: 0px;
    border-radius: 0px;

    z-index: 25;
}
```

While the first empty line of the following would be invalid:

```css
.navbar {

    border: 0px;
    border-radius: 0px;

    z-index: 25;
}
```

> Which version of stylelint are you using?

7.1.0

> How are you running stylelint: CLI, PostCSS plugin, Node API?

CLI using grunt-stylelint and stylelint-scss on NPM. I'm also using linter-stylelint on Atom to lint as I go.

> Does your issue relate to non-standard syntax (e.g. SCSS, nesting, etc.)?

The request itself does not, but I am using SCSS for my syntax.

> What did you expect to happen?

To be able to support the code sample defined above.

> What actually happened (e.g. what warnings or errors you are getting)?

A rule for this couldn't be located.

**Comment by evilebottnawi** [comment]
@heatwaveo8 please format you issue as template, thanks :smile:

**Comment by jeddy3** [comment]
@heatwaveo8 You can achieve your desired code style with the likes of `declaration-empty-line-before`. Don't worry, it's not just you and other people have be caught out by this before. As such, I've [a pending PR](https://github.com/stylelint/stylelint/pull/1850) to update our documentation to explain things a bit more. Have a look through that PR and hopefully things will be clearer.

As @evilebottnawi said, please use the issue template in the future as it helps us understand the issue quicker.

**Comment by heatwaveo8** [comment]
Thanks @jeddy3 for your help. I am still having trouble finding the right solution for this. You can take a look at my revised comment above and see a better code snippet that clarifies my expectations.

At this URL:
http://stylelint.io/user-guide/rules/declaration-empty-line-before/#first-nested

I have found an example that seemed to cover this, but since it is dependent on an "always" or "never" setup, an error is always appearing in my use case with that rule.

Also I apologize for the inconvenience, a lot of the reqs for describing the issue appeared to relate to bugs rather than requests. Perhaps grouping what questions need to be answered for each or clarifying that requests need them too would be of help.

**Comment by jeddy3** [comment]
> I am still having trouble finding the right solution for this

Oh, I see... you *sometimes* separate your declarations with an empty line:

```css
.navbar {
    border: 0px;
    border-radius: 0px;

    z-index: 25;
}
```

What we need is an option to ignore declarations that *come after* other declarations e.g.

```js
"declaration-empty-line-before": ["always", {
  except: [
    "first-nested",
  ],
  ignore: [
    "after-declaration",
  ],
}],
```

Would allow things like:

```css
a {
    border: 0px;
    border-radius: 0px;
    z-index: 25;
}

a {
    border: 0px;
    border-radius: 0px;

    z-index: 25;
}

a {
    --custom-prop: value;

    border: 0px;
    border-radius: 0px;

    z-index: 25;
}

a {
    @custom-at-rule { }

    border: 0px;
    border-radius: 0px;

    z-index: 25;
}
```

But disallow theses:

```css
a {

    border: 0px;
    border-radius: 0px;
    z-index: 25;
}

a {

    border: 0px;
    border-radius: 0px;

    z-index: 25;
}

a {
    --custom-prop: value;
    border: 0px;
    border-radius: 0px;

    z-index: 25;
}

a {
    @custom-at-rule { }
    border: 0px;
    border-radius: 0px;

    z-index: 25;
}
```

@heatwaveo8 Would you like to contribute this option? It would be a great first contribution as it's a small self-contained piece of work. There is already an `ignore: ["after-comment"]` optional secondary option and so it'll just be the case of duplicating and tweaking that logic, tests and docs i.e. [here](https://github.com/stylelint/stylelint/blob/master/src/rules/declaration-empty-line-before/index.js#L37), [here](https://github.com/stylelint/stylelint/blob/master/src/rules/declaration-empty-line-before/index.js#L51-L58), [here](https://github.com/stylelint/stylelint/blob/master/src/rules/declaration-empty-line-before/index.js#L51-L58) and [here](https://github.com/stylelint/stylelint/blob/master/src/rules/declaration-empty-line-before/README.md#ignore-after-comment-inside-single-line-block). You'll need to add a [changelog](https://github.com/stylelint/stylelint/blob/master/CHANGELOG.md) item too: "Added: `ignore: ["after-declaration"]` option to `declaration-empty-line-before`.

FYI, there is also a [plugin proposal](https://github.com/stylelint/stylelint/issues/1575) that would allow you to control which *properties* should be preceded by empty lines and which shouldn't. This is a far more involved piece of work though, but perhaps something you'd be interested in as well later down the line.

> I checked the documentation changes as well and my use case didn't work for that, either. I understand the idea behind "after" referring to the following character

There are no "after" rules, only "before" rules. Hopefully the above (i.e. the proposed addition of the `ignore: ["after-declaration"]` option) has helped clarify how these `*-empty-line-before` rules work.

**Comment by heatwaveo8** [comment]
Right, that's exactly the use case. I think having ignore options for the custom property and custom rule (and others?) would be good as well if they're not already there, to provide flexibility.

Sure I'd be happy to give it a go. Thanks for the references, I'll go through them.

**Comment by jeddy3** [comment]
> Sure I'd be happy to give it a go. 

Fantastic! Good luck.

> I think having ignore options for the custom property and custom rule (and others?) would be good as well if they're not already there, to provide flexibility.

We tend to only add functionality after there has been a real-world request for it. Sometimes adding empty lines between *custom properties* seems like an unusual use-case compared to between declarations. So, I reckon we just focus on adding the option to `declaration-empty-line-before` for now.

**Comment by davidtheclark** [comment]
Closed by 5710a4a09ee0d3da3954701c5702a8ebcfbcba4b.
